debezium-core/src/main/java/io/debezium/function/BufferedBlockingConsumer.java

```diff
@@ -42,27 +43,31 @@ public default void flush() throws InterruptedException {
      * and buffer the latest.
      * <p>
      * The resulting consumer is not threadsafe.
-     * 
+     *
      * @param delegate the delegate to which values should be flushed; may not be null
      * @return the blocking consumer that buffers a single value at a time; never null
      */
     public static <T> BufferedBlockingConsumer<T> bufferLast(BlockingConsumer<T> delegate) {
         return new BufferedBlockingConsumer<T>() {
-            private T last;
+
+            private final AtomicReference<T> last = new AtomicReference<>();
 
             @Override
             public void accept(T t) throws InterruptedException {
-                if (last != null) delegate.accept(last);
-                last = t;
+                T previous = last.getAndSet(t);
+                if (previous != null) {
+                    delegate.accept(previous);
+                }
             }
 
             @Override
-            public void flush(Function<T, T> function) throws InterruptedException {
-                if (last != null) {
+            public synchronized void flush(Function<T, T> function) throws InterruptedException {
+                T previous = last.get();
+                if (previous != null) {
                     try {
-                        delegate.accept(function.apply(last));
+                        delegate.accept(function.apply(previous));
                     } finally {
-                        last = null;
+                        last.set(null);
```

Comment:
Is it possible to run `accept` and `flush` in parallel? If yes then this code will overwrite the value set in `accept` when order of execution is unfortunate.

Reply:
Yeah, it may be possible unless we explicitly change the contract and make sure MongoDB isn't doing this. 
But what if this method were to do the following:

```
T previous = last.getAndSet(null);
if (previous != null) {
    delegate.accept(function.apply(previous));
}
```

In the single-threaded case, there is obviously no concurrency. If multiple threads are concurrently calling `accept(T)`, then this class would work except when one thread get preempted part-way through and another thread calls `accept(T)` and reorders the responses. The same is true if `accept(T)` and `flush(...)` are called concurrently.

However, I'm not sure this matters in the multi-threaded case, right? If there are multiple threads, how can you expect to maintain the order of records from those threads? The records from any single thread will be forwarded in the same order, and IMO that's about as much as we can expect.

And, if you're using this with multiple threads, it seems like it's the responsibility of the caller(s) to ensure that `flush(...)` is called *_after_* all of the `accept(T)` methods have been called. Seems like we need to make sure that MongoDB is doing that, and it might be worth documenting this a bit more.

Reply:
> And, if you're using this with multiple threads, it seems like it's the responsibility of the caller(s) to ensure that flush(...) is called after all of the accept(T) methods have been called. 

Yes, that was my thinking as well and I'd say it is like this in the current MongoDB connector code. `flush()` is only called after the latch is passed, i.e. all the threads have been doing their work.

Reply:
I wasn't concerned about ordering of messages emitted by different threads, and I don't think `synchronized` vs. this proposal does make a difference in this regard. My concern was about the synchronization of all the `accept()` calls by all the threads. I.e. I'm afraid parallelism of these threads is largely hindered by forcing them through this synchronization. That's what I seeked to avoid by using the atomic reference.

Reply:
So, let's document the limitation on `flush` - soething like - this method can be executed if and only if all calls to `accept` has been completed and we can close the issue.

Reply:
Done.